File: backend/app/schemas/product.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, field_validator


class ProductCreate(BaseModel):
    title: str
    description: Optional[str] = None
    price: float
    image_url: Optional[str] = None
    stock: int = 0
    category: Optional[str] = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Title cannot be empty")
        return v.strip()

    @field_validator("price")
    @classmethod
    def price_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("Price must be positive")
        return v

    @field_validator("stock")
    @classmethod
    def stock_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("Stock cannot be negative")
        return v

    @field_validator("image_url")
    @classmethod
    def strip_absolute_url(cls, v: str | None) -> str | None:
        if v is None:
            return v
        for protocol in ("http://", "https://"):
            if v.startswith(protocol):
                slash = v.find("/", len(protocol))
                if slash >= 0:
                    path = v[slash:]
                    if path.startswith("/uploads/"):
                        return path
        return v


class ProductUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    image_url: Optional[str] = None
    stock: Optional[int] = None
    category: Optional[str] = None
    is_active: Optional[bool] = None

    @field_validator("image_url")
    @classmethod
    def strip_absolute_url(cls, v: str | None) -> str | None:
        if v is None:
            return v
        for protocol in ("http://", "https://"):
            if v.startswith(protocol):
                slash = v.find("/", len(protocol))
                if slash >= 0:
                    path = v[slash:]
                    if path.startswith("/uploads/"):
                        return path
        return v


class ProductResponse(BaseModel):
    id: uuid.UUID
    title: str
    description: Optional[str]
    price: float
    image_url: Optional[str]
    stock: int
    category: Optional[str]
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ProductListResponse(BaseModel):
    items: list[ProductResponse]
    total: int
    page: int
    per_page: int

File: frontend/src/services/product.ts
import { api } from "./api";

export interface Product {
  id: string;
  title: string;
  description: string | null;
  price: number;
  image_url: string | null;
  stock: number;
  category: string | null;
  is_active: boolean;
  created_at: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export const productService = {
  async list(params?: {
    category?: string;
    search?: string;
    sort_by?: string;
    page?: number;
    per_page?: number;
  }): Promise<PaginatedResponse<Product>> {
    const q = new URLSearchParams();
    if (params?.category) q.set("category", params.category);
    if (params?.search) q.set("search", params.search);
    if (params?.sort_by) q.set("sort_by", params.sort_by);
    if (params?.page) q.set("page", String(params.page));
    if (params?.per_page) q.set("per_page", String(params.per_page));
    const query = q.toString() ? `?${q}` : "";
    const res = await api.get<{
      items: Product[];
      total: number;
      page: number;
      per_page: number;
    }>(`/products${query}`);
    return {
      items: res.items,
      total: res.total,
      page: res.page,
      per_page: res.per_page,
      total_pages: Math.ceil(res.total / res.per_page),
    };
  },

  async get(id: string): Promise<Product> {
    return api.get<Product>(`/products/${id}`);
  },
};
